src/mongocxx/test/collection.cpp

```diff
@@ -2808,4 +2812,94 @@ TEST_CASE("Ensure that the WriteConcernError 'errInfo' object is propagated", "[
     REQUIRE(contains_err_info);
 }
 
+TEST_CASE("expose writeErrors[].errInfo", "[collection]") {
+    // A helper for checking that an error document is well-formed according to our requirements:
+    auto writeErrors_well_formed = [](const bsoncxx::document::view& reply_view) -> bool {
+        if (!reply_view["writeErrors"]) {
+            return false;
+        }
+
+        const auto& errdoc = reply_view["writeErrors"][0];
+
+        auto error_code = errdoc["code"].get_int32();
+
+        // The code should always be 121 (DocumentValidationFailure):
+        if (121 != error_code) {
+            std::ostringstream os;
+            os << "writeErrors expected to have code 121, but had " << error_code << " instead";
+            throw std::runtime_error(os.str());
+        }
+
+        // We require the "details" field be present:
+        if (!errdoc["errInfo"]["details"]) {
+            throw std::runtime_error("no \"details\" field in \"writeErrors\"");
+        }
+
+        return true;
+    };
+
+    // Set up our test environment:
+    instance::current();
+
+    mongocxx::options::apm apm_opts;
+
+    auto client_opts = test_util::add_test_server_api();
+
+    // Listen to the insertion-failed event: we want to get a copy of the server's
+    // response so that we can compare it to the thrown exception later:
+    apm_opts.on_command_succeeded(
+        [&writeErrors_well_formed](const mongocxx::events::command_succeeded_event& ev) {
+            if (0 != ev.command_name().compare("insert")) {
+                return;
+            }
+
+            REQUIRE(writeErrors_well_formed(ev.reply()));
```

Comment:
To satisfy this requirement:

> Additionally, assert that a CommandSucceededEvent was observed

Please add a check to ensure this callback was called.

For example, one suggestion is to add a boolean flag, and check it at the end of the test:

```cpp
bool insert_succeeded = false;

apm_opts.on_command_succeeded(
        [&writeErrors_well_formed](const mongocxx::events::command_succeeded_event& ev) {
            if (0 != ev.command_name().compare("insert")) {
                return;
            }
            insert_succeeded = true;

            REQUIRE(writeErrors_well_formed(ev.reply()));
}
/* ... */
SECTION("cause a type violation on insert") {
        /* ... */
        REQUIRE(insert_succeeded);
}
```


If the command resulted in a different error producing a CommandFailedEvent, this callback may not be called.

Reply:
+1 I see what you are asking for now-- done!

Reply:
https://github.com/mongodb/mongo-cxx-driver/pull/846/commits/0d0118245cedebb421cf4ce4ee36753fd8eba3ca#diff-481e7ba83bde51981cec950f1b584f37f027c4cb1ee9806d0cfbb73b90483153R2820 adds `insert_succeeded` to the `writeErrors_well_formed` function. The suggestion is to add it to the callback of `on_command_succeeded`.

`writeErrors_well_formed` is called in both the callback to `on_command_succeeded` and in the catch block on L2896.

Please move the `insert_succeeded = true` from `writeErrors_well_formed` to the callback of `on_command_succeeded`.